README.md
# ImP

![npm (scoped)](https://img.shields.io/npm/v/@mischback/imp?style=flat)
[![install size](https://packagephobia.com/badge?p=@mischback/imp)](https://packagephobia.com/result?p=@mischback/imp)

![GitHub package.json version (development)](https://img.shields.io/github/package-json/v/mischback/imp/development?style=flat)
![GitHub branch checks state](https://img.shields.io/github/workflow/status/mischback/imp/CI%20default%20branch?style=flat&logo=github)
[![Coverage Status](https://coveralls.io/repos/github/Mischback/imp/badge.svg)](https://coveralls.io/github/Mischback/imp)
[![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg?style=flat&logo=prettier)](https://github.com/prettier/prettier)
![GitHub License](https://img.shields.io/github/license/mischback/imp?style=flat)

**ImP** is the abbreviation of _image processor_. Basically it is a wrapper
around the wonderful [Sharp](https://github.com/lovell/sharp) module, providing
an easy to use interface to make Sharp usable in a front-end development
workflow.

## Installation

Just install _ImP_ from **npm**:

```bash
npm install --save-dev @mischback/imp
```

Most likely you will want to install it as a development dependency for internal
usage (see below for _Security Considerations_).

## Usage

After installation, _ImP_ is available using the following command:

```bash
npx imp
```

Please note, that a configuration file is required, see below for configuration
options and command line parameters.

A more sophisticated usage example is provided in the
[advanced usage guide](https://github.com/Mischback/imp/blob/development/docs/advanced-usage.md).

## Configuration

_ImP_ is configured by its configuration file and command line parameters.
Providing a configuration file (or the corresponding key in `package.json`) is
mandatory, as the `TargetConfig` and `FormatConfig` can not be specified by
command line.

The in-depth description of the configuration file is provided in
[the configuration guide](https://github.com/Mischback/imp/blob/development/docs/configuration.md).

### Configuration File

_ImP_ uses [cosmiconfig](https://github.com/davidtheclark/cosmiconfig) to
retrieve and read the configuration file.

The following locations are searched:

- key "imp" in `package.json`
- an extensionless "rc file", in YAML or JSON format: `.imprc`
- an "rc file" with extension: `.imprc.json`, `.imprc.yaml`, `.imprc.yml`, `.imprc.js` or `.imprc.cjs`
- a `.config.js` or `.config.cjs` CommonJS module

The listed filenames/locations are checked in the current working directory and
then the search is continued upwards (see [cosmiconfig's README](https://github.com/davidtheclark/cosmiconfig/README.md) for further details).

### Configuration by Command Line

Additionally, _ImP_ accepts the following command line parameters:

- `--debug`, `-d`: Activate the debug mode; in this mode, cosmiconfig's caching is cleared and additional log messages are emitted;
- `--configFile`, `-c`: Specify another name/location for the configuration file to be used;
- `--inputFile`, `-i`: Specify the name/location of the image to be processed. This option may be specified multiple times;
- `--outputDir`, `-o`: Specify the directory to write the processed images to;
- `--quiet`, `-q`: Suppress all output.

Please note that providing `--inputFile` and `--outputDir` through
command line parameters will overwrite the corresponding parameter in the
configuration file.

## Security Considerations

_ImP_ was developed without any special regard to security. It is assumed to be
used in a safe environment, meaning that input and output are under strict of
the user.

You should not expose _ImP_ publicly, at least not without some wrapper that
does perform sanitarization of any user input.

## What about this crazy install size?

_ImP_ is a wrapper around Sharp, which is in fact a very capable interface to
[libvips](https://github.com/libvips/libvips) itsself.

The install size actually reflects the installation of _libvips_.

## Contributing

Issues, pull requests and feature requests are welcome. Just use the project's
[issue tracker](https://github.com/mischback/imp/issues).

_ImP_ is implemented in TypeScript and compiled/transpiled to actual JavaScript
on release.

More details about the internal guidelines regarding code formatting, linting,
testing and releasing are provided in `docs/`.

## License

[MIT](https://choosealicense.com/licenses/MIT)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [2.0.0] - 2022-11-18

- **Dropped** support for NodeJS 12.1 (not longer supported by ``sharp``)
- Dependency of "sharp" was bumped to 0.31.2
- Dependency of "CosmicConfig" was bumped to 7.1.0
- Modified the configuration for ``dependabot`` as per
  https://github.com/Mischback/development-meta/issues/2

## [1.7.0] - 2022-08-09

- Dependency of "sharp" was bumped to 0.30.7

## [1.6.0] - 2022-05-24

- Dependency of "tslog" was bumped to 3.3.3
- Dependency of "minimalist" was bumped to 1.2.6
- Dependecy of "sharp" was bumped to 0.30.5

## [1.4.0] - 2022-02-02

### Added

- Dependency of "sharp" was bumped from 0.29.3 to 0.30.0
  Internally, `sharp` changed some names of options. As these are not
  directly interfaced from _ImP_ but merely handed over to _sharp_, this
  does not affect _ImP_ directly.

## [1.3.0] - 2022-01-04

### Added

- Dependecy of "tslog" was bumped from 3.3.0 to 3.3.1

## [1.2.0] - 2021-11-30

### Added

- Dependency of "tslog" was bumped from 3.2.2 to 3.3.0

## [1.1.0] - 2021-11-15

### Added

- Dependency of "sharp" was bumped from 0.29.2 to 0.29.3
- Improve code quality of Github Actions workflows

## [1.0.0] - 2021-10-29

### Initial release


src/main.spec.ts
// SPDX-FileCopyrightText: 2022 Mischback
// SPDX-License-Identifier: MIT
// SPDX-FileType: SOURCE

/* test specific imports */
import { beforeAll, describe, expect, it, jest } from "@jest/globals";

/* mock library imports */
jest.mock("./lib/configure");
jest.mock("./lib/logging");
jest.mock("./lib/sharprunner");

/* import the subject under test (SUT) */
import { impMain } from "./main";

/* additional imports */
import { getConfig, ImpConfig, ImpConfigureError } from "./lib/configure";
import {
  applyDebugConfiguration,
  applyUserConfiguration,
  logger,
  suppressLogOutput,
} from "./lib/logging";
import { processImageList, SharpRunnerError } from "./lib/sharprunner";

/* Run these before actually starting the test suite */
beforeAll(() => {
  /* The test subject relies on "tslog" to provide log messages.
   * For running the test-suite, actually printing the log messages to the
   * console is unwanted, so the output is suppressed.
   */
  logger.setSettings({
    suppressStdOutput: true,
  });
});

describe("impMain()...", () => {
  it("...attaches SIGINT handler", () => {
    /* define the parameter */
    const testArgv = ["doen't", "matter"];

    /* setup mocks and spies */
    (getConfig as jest.Mock).mockRejectedValue(new ImpConfigureError("foo"));
    const processSpy = jest.spyOn(process, "on");

    /* make the assertions */
    return impMain(testArgv).catch((err: Error) => {
      expect(err).toBe(78);
      expect(processSpy).toHaveBeenCalledTimes(1);
    });
  });

  it("...applies quiet mode", () => {
    /* define the parameter */
    const testArgv = ["doen't", "matter", "-q"];

    /* setup mocks and spies */
    (getConfig as jest.Mock).mockRejectedValue(new ImpConfigureError("foo"));
    (suppressLogOutput as jest.Mock).mockImplementation();

    /* make the assertions */
    return impMain(testArgv).catch((err: Error) => {
      expect(err).toBe(78);
      expect(suppressLogOutput).toHaveBeenCalledTimes(1);
    });
  });

  it("...applies debug mode", () => {
    /* define the parameter */
    const testArgv = ["doen't", "matter", "-d"];

    /* setup mocks and spies */
    (getConfig as jest.Mock).mockRejectedValue(new ImpConfigureError("foo"));
    (applyDebugConfiguration as jest.Mock).mockImplementation();

    /* make the assertions */
    return impMain(testArgv).catch((err: Error) => {
      expect(err).toBe(78);
      expect(applyDebugConfiguration).toHaveBeenCalledTimes(1);
    });
  });

  it("...overrides quiet mode with debug mode", () => {
    /* define the parameter */
    const testArgv = ["doen't", "matter", "-q", "-d"];

    /* setup mocks and spies */
    (getConfig as jest.Mock).mockRejectedValue(new ImpConfigureError("foo"));
    (suppressLogOutput as jest.Mock).mockImplementation();
    (applyDebugConfiguration as jest.Mock).mockImplementation();

    /* make the assertions */
    return impMain(testArgv).catch((err: Error) => {
      expect(err).toBe(78);
      expect(suppressLogOutput).toHaveBeenCalledTimes(1);
      expect(applyDebugConfiguration).toHaveBeenCalledTimes(1);
    });
  });

  it("...applies user-provided logging configuration", () => {
    /* define the parameter */
    const testArgv = ["doen't", "matter"];

    /* setup mocks and spies */
    (getConfig as jest.Mock).mockResolvedValue({
      loggingOptions: {},
    } as ImpConfig);
    (processImageList as jest.Mock).mockRejectedValue(
      new SharpRunnerError("foo")
    );
    (applyUserConfiguration as jest.Mock).mockImplementation();

    /* make the assertions */
    return impMain(testArgv).catch((err: Error) => {
      expect(err).toBe(65);
      expect(applyUserConfiguration).toHaveBeenCalledTimes(1);
    });
  });

  it("...returns the expected exit code 70 for unexpected errors", () => {
    /* define the parameter */
    const testArgv = ["doen't", "matter"];

    /* setup mocks and spies */
    (getConfig as jest.Mock).mockRejectedValue(new Error("foo"));
    /* make the assertions */
    return impMain(testArgv).catch((err: Error) => {
      expect(err).toBe(70);
    });
  });

  it("...correctly terminates with exit code 0 on success", () => {
    /* define the parameter */
    const testArgv = ["doen't", "matter"];

    /* setup mocks and spies */
    (getConfig as jest.Mock).mockResolvedValue({
      loggingOptions: {},
    } as ImpConfig);
    (processImageList as jest.Mock).mockResolvedValue(1337);

    /* make the assertions */
    return impMain(testArgv)
      .then((retVal) => {
        expect(retVal).toBe(0);
      })
      .catch(() => {
        expect(1).toBe(2);
      });
  });
});


src/index.spec.ts
// SPDX-License-Identifier: MIT

/**
 * This file originated by the repository template, that was used to create this
 * repository.
 * It will be removed or adjusted, once there are other tests to be run.
 */

/* Test specific imports */
import { describe, expect, it } from "@jest/globals";

describe("This is just a generic example...", () => {
  it("...to verify that tests can fail", () => {
    expect(1).toBe(2);
  });

  it.only("...to verify that tests can pass", () => {
    expect(1).toBe(1);
  });
});


src/lib/logging.spec.ts
// SPDX-FileCopyrightText: 2022 Mischback
// SPDX-License-Identifier: MIT
// SPDX-FileType: SOURCE

/* test specific imports */
import { beforeAll, describe, expect, it, jest } from "@jest/globals";

/* mock library imports */
// INSERT LIB IMPORTS HERE!

/* import the subject under test (SUT) */
import {
  applyDebugConfiguration,
  applyLoggingConfiguration,
  applyUserConfiguration,
  logger,
  suppressLogOutput,
} from "./logging";

/* additional imports */
import { ISettingsParam, ISettings } from "tslog";

/* Run these before actually starting the test suite */
beforeAll(() => {
  /* The test subject relies on "tslog" to provide log messages.
   * For running the test-suite, actually printing the log messages to the
   * console is unwanted, so the output is suppressed.
   */
  logger.setSettings({
    suppressStdOutput: true,
  });
});

describe("applyDebugConfiguration()...", () => {
  it("...applies the provided default debug config", () => {
    /* This is actually the provided logging configuration for the debug mode.
     * It is copy/pasted from "logging.ts", because it is not exported. However,
     * this means that this test will fail, if/when this default config is
     * modified.
     */
    const debugLoggingConfiguration: ISettingsParam = {
      minLevel: "debug",
      displayFunctionName: true,
      displayFilePath: "hideNodeModulesOnly",
      printLogMessageInNewLine: true,
      suppressStdOutput: false,
    };

    /* setup mocks and spies */
    const spyLoggerSetSettings = jest
      .spyOn(logger, "setSettings")
      .mockImplementationOnce((_settings) => {
        return {} as ISettings;
      });
    const spyLoggerDebug = jest.spyOn(logger, "debug");

    /* make the assertions */
    expect(applyDebugConfiguration()).toBe(undefined);
    expect(spyLoggerSetSettings).toHaveBeenCalledTimes(1);
    expect(spyLoggerSetSettings).toHaveBeenCalledWith(
      debugLoggingConfiguration
    );
    expect(spyLoggerDebug).toHaveBeenCalledTimes(1);
  });
});

describe("applyLoggingConfiguration()...", () => {
  it("...applies the configuration as provided by parameter", () => {
    /* define the parameter */
    const testConfiguration: ISettingsParam = {
      name: "TESTCASE",
      minLevel: "error",
      printLogMessageInNewLine: false,
    };

    /* setup mocks and spies */
    const spyLoggerSetSettings = jest.spyOn(logger, "setSettings");

    /* make the assertions */
    expect(applyLoggingConfiguration(testConfiguration)).toBe(undefined);
    expect(spyLoggerSetSettings).toHaveBeenCalledTimes(1);
    expect(spyLoggerSetSettings).toHaveBeenCalledWith(testConfiguration);
  });
});

describe("applyUserConfiguration()...", () => {
  it("...applies the provided user config", () => {
    /* This is actually the provided logging configuration for the debug mode.
     * It is copy/pasted from "logging.ts", because it is not exported. However,
     * this means that this test will fail, if/when this default config is
     * modified.
     */
    const testConfiguration: ISettingsParam = {
      name: "TESTCASE",
      minLevel: "error",
      printLogMessageInNewLine: false,
    };

    /* setup mocks and spies */
    const spyLoggerSetSettings = jest
      .spyOn(logger, "setSettings")
      .mockImplementationOnce((_settings) => {
        return {} as ISettings;
      });
    const spyLoggerDebug = jest.spyOn(logger, "debug");

    /* make the assertions */
    expect(applyUserConfiguration(testConfiguration, false)).toBe(undefined);
    expect(spyLoggerSetSettings).toHaveBeenCalledTimes(2);
    expect(spyLoggerSetSettings).toHaveBeenCalledWith(testConfiguration);
    expect(spyLoggerDebug).toHaveBeenCalledTimes(0);
  });

  it("...applies the provided user config and keeps debugMode", () => {
    /* This is actually the provided logging configuration for the debug mode.
     * It is copy/pasted from "logging.ts", because it is not exported. However,
     * this means that this test will fail, if/when this default config is
     * modified.
     */
    const testConfiguration: ISettingsParam = {
      name: "TESTCASE",
      minLevel: "error",
      printLogMessageInNewLine: false,
    };

    /* setup mocks and spies */
    const spyLoggerSetSettings = jest
      .spyOn(logger, "setSettings")
      .mockImplementationOnce((_settings) => {
        return {} as ISettings;
      });
    const spyLoggerDebug = jest.spyOn(logger, "debug");

    /* make the assertions */
    expect(applyUserConfiguration(testConfiguration, true)).toBe(undefined);
    expect(spyLoggerSetSettings).toHaveBeenCalledTimes(3);
    expect(spyLoggerSetSettings).toHaveBeenCalledWith(testConfiguration);
    expect(spyLoggerDebug).toHaveBeenCalledTimes(0);
  });
});

describe("suppressLogOutput()...", () => {
  it("...applies the argument to suppress output", () => {
    /* define the parameter */
    const testConfiguration: ISettingsParam = { suppressStdOutput: true };

    /* setup mocks and spies */
    const spyLoggerSetSettings = jest.spyOn(logger, "setSettings");

    /* make the assertions */
    expect(suppressLogOutput()).toBe(undefined);
    expect(spyLoggerSetSettings).toHaveBeenCalledTimes(1);
    expect(spyLoggerSetSettings).toHaveBeenCalledWith(testConfiguration);
  });
});


src/lib/logging.ts
// SPDX-FileCopyrightText: 2022 Mischback
// SPDX-License-Identifier: MIT
// SPDX-FileType: SOURCE

/* Library imports */
import { ISettingsParam, Logger } from "tslog";

/**
 * The module's logging configuration for debug mode.
 *
 * This configuration is provided for development. The module's user may provide
 * their own logging configuration in their configuration file, which
 * effectively overrides this one.
 *
 * See documentation of "tslog" for a list of available options.
 *
 * @see ISettingsParam
 */
const debugLoggingConfiguration: ISettingsParam = {
  minLevel: "debug",
  displayFunctionName: true,
  displayFilePath: "hideNodeModulesOnly",
  printLogMessageInNewLine: true,
  suppressStdOutput: false,
};

/**
 * The module's default logging configuration.
 *
 * This configuration is provided for development. The module's user may provide
 * their own logging configuration in their configuration file, which
 * effectively overrides this one.
 *
 * See documentation of "tslog" for a list of available options.
 *
 * @see ISettingsParam
 */
const defaultLoggingConfiguration: ISettingsParam = {
  name: "ImP",
  displayLoggerName: true,
  minLevel: "info",
  displayFunctionName: false,
  displayFilePath: "hidden",
};

/**
 * The modules default logger.
 *
 * It will be initialized with sane default configuration. The configuration may
 * be overridden with a user-provided logging configuration.
 *
 * See documentation of "tslog" for a list of available options.
 *
 * @see defaultLoggingConfiguration
 * @see ISettingsParam
 */
export const logger = new Logger(defaultLoggingConfiguration);

/**
 * Apply a logging configuration to the logger.
 *
 * @param loggingConfig - The config to apply.
 *
 * See documentation of "tslog" for a list of available options.
 *
 * @see ISettingsParam
 */
export function applyLoggingConfiguration(loggingConfig: ISettingsParam): void {
  logger.setSettings(loggingConfig);
}

/**
 * Apply the default logging configuration for debug mode.
 */
export function applyDebugConfiguration(): void {
  applyLoggingConfiguration(debugLoggingConfiguration);
  logger.debug("Activated debug mode...");
}

/**
 * Apply a user-provided logging configuration.
 *
 * @param loggingConfig - The configuration to apply, as provided by a
 *                        configuration file.
 * @param debug - A flag to indicate, if the module was already running in
 *                debug mode.
 *
 * Before the user-provided logging configuration is applied, every value that
 * may be modified by the modules default logging configuration or default
 * debugging configuration is set to its default value as provided by "tslog".
 *
 * After applying the user-provided configuration, the debug-flag is evaluated
 * and the minLevel is adjusted to keep running with log level "debug".
 *
 * @see applyLoggingConfiguration
 * @see resetLoggingConfiguration
 * @see ISettingsParam
 */
export function applyUserConfiguration(
  loggingConfig: ISettingsParam,
  debug: boolean
): void {
  resetLoggingConfiguration();
  applyLoggingConfiguration(loggingConfig);

  if (debug === true) logger.setSettings({ minLevel: "debug" });
}

/**
 * Reset all tslog options back to their defaults
 *
 * This function will only reset those values, that are modified by the module's
 * pre-defined logging configurations.
 *
 * @see defaultLoggingConfiguration
 * @see debugLoggingConfiguration
 */
function resetLoggingConfiguration(): void {
  logger.setSettings({
    displayFunctionName: true,
    displayLoggerName: true,
    displayFilePath: "hideNodeModulesOnly",
    minLevel: "silly",
    printLogMessageInNewLine: false,
    suppressStdOutput: false,
  });
}

/**
 * Suppress all logging outputs.
 *
 * This function is used to provide the quiet mode.
 *
 * Please note: If there is a user-provided logging-configuration, this may be
 * overridden. Basically this means: if the user want to suppress logging output
 * AND provide his own logging configuration, he will have to include the option
 * "suppressStdOutput" in his configuration.
 */
export function suppressLogOutput(): void {
  logger.setSettings({ suppressStdOutput: true });
}


src/lib/sharprunner.ts
// SPDX-FileCopyrightText: 2022 Mischback
// SPDX-License-Identifier: MIT
// SPDX-FileType: SOURCE

/* library imports */
import { createReadStream } from "fs";
import { basename, extname, join } from "path";
import sharp = require("sharp");

/* internal imports */
import {
  FormatConfig,
  ImpConfig,
  TargetConfig,
  TargetConfigItem,
  TargetFormat,
} from "./configure";
import { ImpError } from "./errors";
import { logger } from "./logging";

/* *** INTERNAL CONSTANTS *** */

export class SharpRunnerError extends ImpError {
  constructor(message: string) {
    super(message);
  }
}

class SharpRunnerProcessError extends SharpRunnerError {
  constructor(message: string) {
    super(message);
  }
}

class SharpRunnerCreatePipeError extends SharpRunnerError {
  constructor(message: string) {
    super(message);
  }
}

class SharpRunnerCreatePipeFormatError extends SharpRunnerCreatePipeError {
  constructor(message: string) {
    super(message);
  }
}

class SharpRunnerCreatePipeModeError extends SharpRunnerCreatePipeError {
  constructor(message: string) {
    super(message);
  }
}

/**
 * Provides the actual file extensions for Sharp's accepted output formats.
 *
 * @see TargetFormat
 */
const TargetFormatExtensions: { [index: string]: string } = {
  avif: ".avif",
  gif: ".gif",
  heif: ".heif",
  jpeg: ".jpg",
  jpg: ".jpg",
  png: ".png",
  tiff: ".tiff",
  webp: ".webp",
};

/**
 * This class is the core of the module. It provides all necessary means to
 * process one input file into one or several output files.
 */
export class SharpRunner {
  _inputFile: string;
  _outputDir: string;
  _targets: TargetConfig;
  _formatOptions: FormatConfig;

  __sharpPipeEntry: sharp.Sharp | undefined;
  _fileBasename: string;

  /**
   * Create a new instance of SharpRunner.
   *
   * @param inputFile - Path/filename of the file to process
   * @param config - An instance of ImpConfig
   * @returns - An instance of SharpRunner
   *
   * The constructor will split the provided ImpConfig instance into its parts
   * and setup the internal state of the object.
   */
  constructor(inputFile: string, config: ImpConfig) {
    this._inputFile = inputFile;
    this._outputDir = config.outputDir;
    this._targets = config.targets;
    this._formatOptions = config.formatOptions;

    this.__sharpPipeEntry = undefined;
    this._fileBasename = basename(this._inputFile, extname(this._inputFile));
  }

  /**
   * Trigger the actual processing of the input image.
   *
   * @returns Number of processed pipes = number of created images
   *
   * This is the actual public interface of the class (beside its constructor).
   */
  process(): Promise<number> {
    return new Promise((resolve, reject) => {
      this._buildPipes()
        .then((sharpPipes) => {
          return this._processPipes(sharpPipes);
        })
        .then((numberOfProcessedPipes) => {
          return resolve(numberOfProcessedPipes);
        })
        .catch((err: Error) => {
          if (err instanceof SharpRunnerError) return reject(err);

          logger.debug(err);
          return reject(
            new SharpRunnerProcessError(
              `Unexpected error while processing ${this._inputFile}`
            )
          );
        });
    });
  }

  /**
   * Build the required pipes as specified by targets.
   *
   * @returns - A Promise, resolving to a list of sharp.Sharp instances
   *
   * The actual pipes are built by _createPipe(). This method basically loops
   * through the targets: TargetConfig and pushes every result of _createPipe()
   * into a list of sharpPipes: sharp.Sharp[].
   *
   * @see _createPipe
   * @see TargetConfig
   */
  _buildPipes(): Promise<sharp.Sharp[]> {
    return new Promise((resolve, reject) => {
      const sharpPipes: sharp.Sharp[] = [];

      Object.keys(this._targets).forEach((targetKey) => {
        const target = this._targets[targetKey];

        target?.formats.forEach((targetFormat) => {
          try {
            sharpPipes.push(this._createPipe(target, targetFormat));
          } catch (err) {
            if (err instanceof SharpRunnerCreatePipeFormatError) {
              logger.debug(err);
              logger.warn(
                `Found an unknown target format: "${targetFormat}". Skipping!`
              );
            } else {
              return reject(err);
            }
          }
        });
      });

      logger.debug("Completed building of pipes. Resolving!");
      return resolve(sharpPipes);
    });
  }

  /**
   * Create a single pipe as instance of sharp.Sharp.
   *
   * @param target - An instance of TargetConfigItem
   * @param targetFormat - The format for the output file
   * @returns - An instance of sharp.Sharp
   *
   * @see _buildPipes
   * @see TargetConfigItem
   * @see TargetFormat
   */
  _createPipe(
    target: TargetConfigItem,
    targetFormat: TargetFormat
  ): sharp.Sharp {
    let targetSharpFormat: keyof sharp.FormatEnum;
    if (targetFormat in sharp.format) {
      targetSharpFormat = targetFormat as keyof sharp.FormatEnum;
    } else {
      throw new SharpRunnerCreatePipeFormatError(
        `Unknown target format "${targetFormat}"`
      );
    }

    const fileBasename =
      target?.filenameSuffix !== undefined
        ? this._fileBasename + target?.filenameSuffix
        : this._fileBasename;

    const newFilename = `${fileBasename}${
      TargetFormatExtensions[targetFormat] as string
    }`;

    const targetFormatOptions =
      this._formatOptions[targetFormat] !== undefined
        ? this._formatOptions[targetFormat]
        : {};

    let pipe = this._sharpPipeEntry.clone();

    switch (target.mode) {
      case "do-not-scale":
        break;
      case "keep-aspect":
        if (target.width !== undefined) {
          pipe = pipe.resize({ width: target.width });
        } else if (target.height !== undefined) {
          pipe = pipe.resize({ height: target.height });
        } else
          throw new SharpRunnerCreatePipeModeError(
            'Mode "keep-aspect" requires either "width" or "height" to be specified'
          );
        break;
      default:
        throw new SharpRunnerCreatePipeModeError("Unknown target mode");
    }

    pipe = pipe.toFormat(targetSharpFormat, targetFormatOptions);

    // The following line is ignored from TypeScript's and eslint's checks,
    // because they find, that the Promise is not fully populated. Indeed, it
    // will be fully populated, but TypeScript can not determine this, because
    // the call to "toFormat()" is performed with a variable filetype.
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    pipe = pipe.toFile(join(this._outputDir, newFilename));

    logger.debug(`Built pipe for "${newFilename}"`);

    return pipe;
  }

  /**
   * Process the pipes by piping the input file into the prepared pipes.
   *
   * @param sharpPipes - A list of sharp.Sharp instances to process
   * @returns - A Promise, resolving to the number of processed pipes / created
   *            output files
   *
   * @see _buildPipes
   * @see _createPipe
   */
  _processPipes(sharpPipes: sharp.Sharp[]): Promise<number> {
    return new Promise((resolve, reject) => {
      if (sharpPipes.length === 0) {
        logger.warn("There are no pipes to process. Aborting!");
        return reject(new SharpRunnerProcessError("No pipes to process"));
      }

      const stream = createReadStream(this._inputFile);

      stream.on("open", () => {
        logger.debug(`Piping "${this._inputFile}" into the sharp pipes...`);
        stream.pipe(this._sharpPipeEntry);
      });

      stream.on("error", (err) => {
        logger.debug(err);
        return reject(
          new SharpRunnerProcessError(
            `Error while reading "${this._inputFile}"`
          )
        );
      });

      Promise.all(sharpPipes)
        .then(() => {
          logger.debug("Completed processing all pipes. Resolving!");
          return resolve(sharpPipes.length);
        })
        .catch((err: Error) => {
          logger.debug(err);
          return reject(
            new SharpRunnerProcessError("Error while processing the pipes")
          );
        });
    });
  }

  /**
   * Get this instance's entry point of the Sharp-based processing pipes.
   *
   * @returns sharp.Sharp instance
   *
   * Basically this implements a singleton pattern to ensure that there is one
   * and only one entry point for sharp's processing.
   *
   * This is an internal implementation detail and not part of the public
   * interface of this class!
   */
  get _sharpPipeEntry() {
    if (this.__sharpPipeEntry === undefined) {
      this.__sharpPipeEntry = sharp({ failOnError: true });
    }

    return this.__sharpPipeEntry;
  }
}

/**
 * Process a list of input files by creating instances of SharpRunner.
 *
 * @param configObject - An instance of ImpConfig
 * @returns - A Promise, resolving to the total number of processed pipes
 *
 * @see SharpRunner
 * @see ImpConfig
 */
export function processImageList(configObject: ImpConfig): Promise<number> {
  return new Promise((resolve, reject) => {
    const imageList: Promise<number>[] = [];

    configObject.inputFiles.forEach((inputFile: string) => {
      imageList.push(new SharpRunner(inputFile, configObject).process());
    });

    Promise.all(imageList)
      .then((retVals) => {
        logger.debug(retVals);
        return resolve(retVals.reduce((a, b) => a + b, 0));
      })
      .catch((err: Error) => {
        return reject(err);
      });
  });
}


src/main.ts
// SPDX-FileCopyrightText: 2022 Mischback
// SPDX-License-Identifier: MIT
// SPDX-FileType: SOURCE

/* library imports */
import { Config } from "stdio/dist/getopt";

/* internal imports */
import { cmdLineOptions, getConfig, ImpConfigureError } from "./lib/configure";
import {
  logger,
  applyDebugConfiguration,
  suppressLogOutput,
  applyUserConfiguration,
} from "./lib/logging";
import { processImageList, SharpRunnerError } from "./lib/sharprunner";

/* *** INTERNAL CONSTANTS *** */
const EXIT_SUCCESS = 0; // sysexits.h: 0 -> successful termination
const EXIT_DATA_ERROR = 65; // sysexits.h: 65 -> input data was incorrect in some way
const EXIT_INTERNAL_ERROR = 70; // sysexits.h: 70 -> internal software error
const EXIT_CONFIG_ERROR = 78; // sysexits.h: 78 -> configuration error
const EXIT_SIGINT = 130; // bash scripting guide: 130 -> terminated by ctrl-c

/* *** TYPE DEFINITIONS *** */
type StdioConfigItem = Exclude<Config, boolean | undefined>;

/**
 * The module's main function. It controls the general flow of execution.
 *
 * @param argv - The argument vector as a list of strings
 * @returns - A Promise, resolving to a number, representing an exit code
 *
 * @see getConfig
 * @see SharpRunner
 * @see processImageList
 */
export function impMain(argv: string[]): Promise<number> {
  return new Promise((resolve, reject) => {
    /* Setting up a handler for SIGINT (Ctrl-C)
     * This handler may be useful for cleaning up before terminating the script.
     * At least it will resolve to the "correct" exit code.
     */
    process.on("SIGINT", () => {
      logger.info("Caught interrupt signal (Ctrl-C). Exiting!");
      return reject(EXIT_SIGINT);
    });

    /* Activate the quiet mode as early as possible
     * This is done without getopt() from stdio, because getopt() will be called
     * later during startup.
     * Please note: if quiet mode and debug mode are activated, debug mode wins.
     */
    const quietKey = (cmdLineOptions.quiet as StdioConfigItem)["key"];
    if (argv.indexOf(`-${quietKey as string}`) > -1) {
      suppressLogOutput();
    }

    /* Activate the debug mode as early as possible
     * This is done without getopt() from stdio, because getopt() will be called
     * later during startup.
     */
    let debugMode = false;
    const debugKey = (cmdLineOptions.debug as StdioConfigItem)["key"];
    if (argv.indexOf(`-${debugKey as string}`) > -1) {
      applyDebugConfiguration();
      debugMode = true;
    }

    /* The actual payload starts here */
    getConfig(argv)
      .then((configObject) => {
        if (configObject.loggingOptions !== undefined) {
          applyUserConfiguration(configObject.loggingOptions, debugMode);
        }

        return Promise.resolve(configObject);
      })
      .then((configObject) => {
        return processImageList(configObject);
      })
      .then((retVal) => {
        logger.info(`Processed ${retVal} pipes!`);
        return resolve(EXIT_SUCCESS);
      })
      .catch((err: Error) => {
        /* handle "known" errors */
        if (err instanceof ImpConfigureError) {
          logger.error(err.message);
          logger.fatal("Could not determine configuration for ImP!");
          return reject(EXIT_CONFIG_ERROR);
        }

        if (err instanceof SharpRunnerError) {
          logger.error(err.message);
          logger.fatal("Could not process images!");
          return reject(EXIT_DATA_ERROR);
        }

        /* general error handler */
        logger.error("Whoops, that was unexpected!");
        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
        logger.error(err.message);
        logger.fatal("An unexpected error occured. Aborting!");
        return reject(EXIT_INTERNAL_ERROR);
      });
  });
}
